File: functions_IPC.R
IPC <- function(x, y) {
  n <- nrow(y)
  
  xxup <- x %*% t(x) + sum(colMeans(x^2))
  xxdown <- sqrt(diag(xxup))
  xxdown <- outer(xxdown, xxdown)
  b0 <- asin(xxup / xxdown)
  b <- Re(b0)
  diag(b) <- 0
  bjsum <- rowSums(b)
  bbsum <- sum(bjsum)
  
  yyup <- y %*% t(y) + sum(colMeans(y^2))
  yydown <- sqrt(diag(yyup))
  yydown <- outer(yydown, yydown)
  rhok0 <- asin(yyup / yydown)
  rhok <- Re(rhok0)
  diag(rhok) <- 0
  rhoksum <- rowSums(rhok)
  rhosum <- sum(rhoksum)
  
  S1 <- mean(rowSums(b * rhok)) / (n - 3)
  S2 <- mean(bjsum * rhoksum) / ((n - 3) * (n - 2))
  S3 <- rhosum * bbsum / (n * (n - 3) * (n - 2) * (n - 1))
  CVM2 <- S1 - 2 * S2 + S3
  
  S1 <- mean(rowSums(b * b)) / (n - 3)
  S2 <- mean(bjsum * bjsum) / ((n - 3) * (n - 2))
  S3 <- bbsum * bbsum / (n * (n - 3) * (n - 2) * (n - 1))
  vx <- S1 - 2 * S2 + S3
  
  S1 <- mean(rowSums(rhok * rhok)) / (n - 3)
  S2 <- mean(rhoksum * rhoksum) / ((n - 3) * (n - 2))
  S3 <- rhosum * rhosum / (n * (n - 3) * (n - 2) * (n - 1))
  vy <- S1 - 2 * S2 + S3
  
  IPC2 <- CVM2 / sqrt(vx * vy)
  ustat <- n * CVM2 / sqrt(2 * vx * vy)
  
  return(list(ustat = ustat, IPC2 = IPC2))
}

File: functions_IPC.py
import numpy as np

def IPC(x, y):
    n = y.shape[0]

    xxup = x @ x.T + np.sum(np.mean(x**2, axis=0))
    xxdown = np.sqrt(np.diag(xxup))
    xxdown = np.outer(xxdown, xxdown)
    b0 = np.arcsin(xxup / xxdown)
    b = np.real(b0)
    np.fill_diagonal(b, 0)
    bjsum = np.sum(b, axis=1)
    bbsum = np.sum(bjsum)

    yyup = y @ y.T + np.sum(np.mean(y**2, axis=0))
    yydown = np.sqrt(np.diag(yyup))
    yydown = np.outer(yydown, yydown)
    rhok0 = np.arcsin(yyup / yydown)
    rhok = np.real(rhok0)
    np.fill_diagonal(rhok, 0)
    rhoksum = np.sum(rhok, axis=1)
    rhosum = np.sum(rhoksum)

    S1 = np.mean(np.sum(b * rhok, axis=1)) / (n - 3)
    S2 = np.mean(bjsum * rhoksum) / ((n - 3) * (n - 2))
    S3 = rhosum * bbsum / (n * (n - 3) * (n - 2) * (n - 1))
    CVM2 = S1 - 2 * S2 + S3

    S1 = np.mean(np.sum(b * b, axis=1)) / (n - 3)
    S2 = np.mean(bjsum * bjsum) / ((n - 3) * (n - 2))
    S3 = bbsum * bbsum / (n * (n - 3) * (n - 2) * (n - 1))
    vx = S1 - 2 * S2 + S3

    S1 = np.mean(np.sum(rhok * rhok, axis=1)) / (n - 3)
    S2 = np.mean(rhoksum * rhoksum) / ((n - 3) * (n - 2))
    S3 = rhosum * rhosum / (n * (n - 3) * (n - 2) * (n - 1))
    vy = S1 - 2 * S2 + S3

    IPC2 = CVM2 / np.sqrt(vx * vy)
    ustat = n * CVM2 / np.sqrt(2 * vx * vy)

    return ustat, IPC2
